File: app/src/main/java/piuk/blockchain/android/util/ExchangeRateFactory.java
package piuk.blockchain.android.util;

import info.blockchain.api.data.TickerItem;
import info.blockchain.api.exchangerates.ExchangeRates;
import info.blockchain.wallet.BlockchainFramework;
import info.blockchain.wallet.api.WalletApi;
import info.blockchain.wallet.exceptions.ApiException;

import java.math.BigInteger;
import java.text.NumberFormat;
import java.util.Locale;
import java.util.TreeMap;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Observable;
import piuk.blockchain.android.data.rxjava.RxBus;
import piuk.blockchain.android.data.rxjava.RxPinning;
import piuk.blockchain.android.data.rxjava.RxUtil;
import piuk.blockchain.android.injection.Injector;
import retrofit2.Response;

/**
 * This class obtains info on the currencies communicated via https://blockchain.info/ticker
 */
public class ExchangeRateFactory {

    private static final String PREF_LAST_KNOWN_BTC_PRICE = "LAST_KNOWN_BTC_VALUE_FOR_CURRENCY_";
    private static final String PREF_LAST_KNOWN_ETH_PRICE = "LAST_KNOWN_ETH_VALUE_FOR_CURRENCY_";

    private final WalletApi walletApi = new WalletApi();
    //Regularly updated ticker data
    private TreeMap<String, TickerItem> btcTickerData;
    private TreeMap<String, TickerItem> ethTickerData;
    private ExchangeRates api;

    @Inject protected PrefsUtil prefsUtil;
    @Inject protected RxBus rxBus;
    private RxPinning rxPinning;

    private static ExchangeRateFactory instance = null;

    private ExchangeRateFactory() {
        Injector.getInstance().getAppComponent().inject(this);
        api = new ExchangeRates(
                BlockchainFramework.getRetrofitExplorerInstance(),
                BlockchainFramework.getRetrofitApiInstance(),
                BlockchainFramework.getApiCode());

        rxPinning = new RxPinning(rxBus);
    }

    public static ExchangeRateFactory getInstance() {
        if (instance == null) {
            instance = new ExchangeRateFactory();
        }

        return instance;
    }

    public Completable updateTickers() {
        return rxPinning.call(() -> getBtcTicker().mergeWith(getEthTicker()));
    }

    public double getLastBtcPrice(String currencyName) {
        if (currencyName.isEmpty()) {
            currencyName = "USD";
        }

        double lastPrice;
        double lastKnown = Double.parseDouble(prefsUtil.getValue(PREF_LAST_KNOWN_BTC_PRICE + currencyName, "0.0"));

        if (btcTickerData == null) {
            lastPrice = lastKnown;
        } else {
            TickerItem tickerItem = getBtcTickerItem(currencyName);
            lastPrice = tickerItem.getLast();

            if (lastPrice > 0.0) {
                prefsUtil.setValue(PREF_LAST_KNOWN_BTC_PRICE + currencyName, Double.toString(lastPrice));
            } else {
                lastPrice = lastKnown;
            }
        }

        return lastPrice;
    }

    public double getLastEthPrice(String currencyName) {
        if (currencyName.isEmpty()) {
            currencyName = "USD";
        }

        double lastPrice;
        double lastKnown = Double.parseDouble(prefsUtil.getValue(PREF_LAST_KNOWN_ETH_PRICE + currencyName, "0.0"));

        if (ethTickerData == null) {
            lastPrice = lastKnown;
        } else {
            TickerItem tickerItem = getEthTickerItem(currencyName);
            lastPrice = tickerItem.getLast();

            if (lastPrice > 0.0) {
                prefsUtil.setValue(PREF_LAST_KNOWN_ETH_PRICE + currencyName, Double.toString(lastPrice));
            } else {
                lastPrice = lastKnown;
            }
        }

        return lastPrice;
    }

    public String getSymbol(String currencyName) {
        if (currencyName.isEmpty()) {
            currencyName = "USD";
        }

        TickerItem tickerItem = getBtcTickerItem(currencyName);
        return tickerItem.getSymbol();
    }

    /**
     * Returns the historic value of a number of Satoshi at a given time in a given currency.
     *
     * @param satoshis     The amount of Satoshi to be converted
     * @param currency     The currency to be converted to as a 3 letter acronym, eg USD, GBP
     * @param timeInMillis The time at which to get the price, in milliseconds since epoch
     * @return A double value
     */
    public Observable<Double> getBtcHistoricPrice(long satoshis, String currency, long timeInMillis) {
        return walletApi.getBtcHistoricPrice(satoshis, currency, timeInMillis)
                .flatMap(responseBody -> parseStringValue(responseBody.string()))
                .compose(RxUtil.applySchedulersToObservable());
    }

    /**
     * Returns the historic value of a number of Wei at a given time in a given currency.
     *
     * @param wei          The amount of Ether to be converted in Wei, ie ETH * 1e18
     * @param currency     The currency to be converted to as a 3 letter acronym, eg USD, GBP
     * @param timeInMillis The time at which to get the price, in milliseconds since epoch
     * @return A double value
     */
    public Observable<Double> getEthHistoricPrice(BigInteger wei, String currency, long timeInMillis) {
        return walletApi.getEthHistoricPrice(wei, currency, timeInMillis)
                .flatMap(responseBody -> parseStringValue(responseBody.string()))
                .compose(RxUtil.applySchedulersToObservable());
    }

    private Observable<Double> parseStringValue(String value) {
        return Observable.fromCallable(() -> {
            // Historic prices are in English format, using Locale.getDefault() will result in
            // a parse exception in some regions
            NumberFormat format = NumberFormat.getInstance(Locale.ENGLISH);
            Number number = format.parse(value);
            return number.doubleValue();
        });
    }

    public String[] getCurrencyLabels() {
        return btcTickerData.keySet().toArray(new String[0]);
    }

    private Completable getBtcTicker() {
        return Completable.fromCallable(() -> {
            Response<TreeMap<String, TickerItem>> call = api.getBtcTickerMap().execute();
            if (call.isSuccessful()) {
                btcTickerData = call.body();
                return Void.TYPE;
            } else {
                throw new ApiException(call.errorBody().string());
            }
        }).compose(RxUtil.applySchedulersToCompletable());
    }

    private Completable getEthTicker() {
        return Completable.fromCallable(() -> {
            Response<TreeMap<String, TickerItem>> call = api.getEthTickerMap().execute();
            if (call.isSuccessful()) {
                ethTickerData = call.body();
                return Void.TYPE;
            } else {
                throw new ApiException(call.errorBody().string());
            }
        }).compose(RxUtil.applySchedulersToCompletable());
    }

    private TickerItem getBtcTickerItem(String currencyName) {
        return btcTickerData.get(currencyName);
    }

    private TickerItem getEthTickerItem(String currencyName) {
        return ethTickerData.get(currencyName);
    }
}

File: app/src/main/java/piuk/blockchain/android/util/ExchangeRateFactory.kt
package piuk.blockchain.android.util

import info.blockchain.wallet.prices.PriceApi
import info.blockchain.wallet.prices.data.PriceDatum
import io.reactivex.Observable
import piuk.blockchain.android.data.currency.CryptoCurrencies
import piuk.blockchain.android.data.rxjava.RxBus
import piuk.blockchain.android.data.rxjava.RxPinning
import piuk.blockchain.android.data.rxjava.RxUtil
import piuk.blockchain.android.data.stores.Optional
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.util.annotations.Mockable
import piuk.blockchain.android.util.helperfunctions.unsafeLazy
import java.math.BigDecimal
import java.math.BigInteger
import java.util.*
import javax.inject.Inject

/**
 * This class obtains info on the currencies communicated via https://blockchain.info/ticker. This
 * is set up like a classical Singleton for compatibility's sake, and would normally be a Kotlin
 * object instead. However, we need this class open for testing, and we (for now) need to access
 * it outside of dependency injection.
 */
@Mockable
class ExchangeRateFactory private constructor() {

    private object Holder {
        val INSTANCE = ExchangeRateFactory()
    }

    private final val priceApi by unsafeLazy { PriceApi() }
    private final val rxPinning: RxPinning
    // Ticker data
    private var btcTickerData: Map<String, PriceDatum>? = null
    private var ethTickerData: Map<String, PriceDatum>? = null

    @Inject final lateinit var prefsUtil: PrefsUtil
    @Inject final lateinit var rxBus: RxBus

    init {
        @Suppress("LeakingThis") // This will be resolved in the future
        Injector.getInstance().appComponent.inject(this)
        rxPinning = RxPinning(rxBus)
    }

    fun updateTickers(): Observable<Map<String, PriceDatum>> = rxPinning.call<Map<String, PriceDatum>> {
        getBtcTicker().mergeWith(getEthTicker())
    }

    fun getLastBtcPrice(currencyName: String) = getLastPrice(currencyName.toUpperCase(), CryptoCurrencies.BTC)

    fun getLastEthPrice(currencyName: String) = getLastPrice(currencyName.toUpperCase(), CryptoCurrencies.ETHER)

    @Deprecated(
            "use MonetaryUtil.getCurrencySymbol, as this method doesn't allow the passing of a locale",
            replaceWith = ReplaceWith(
                    "monetaryUtil.getCurrencySymbol(currencyName, Locale.getDefault())",
                    "piuk.blockchain.android.util.MonetaryUtil"
            )
    )
    fun getSymbol(currencyName: String): String {
        var currency = currencyName
        if (currency.isEmpty()) {
            currency = "USD"
        }
        return Currency.getInstance(currency).getSymbol(Locale.getDefault())
    }

    fun getCurrencyLabels(): Array<String> = btcTickerData!!.keys.toTypedArray()

    /**
     * Returns the historic value of a number of Satoshi at a given time in a given currency.
     *
     * @param satoshis     The amount of Satoshi to be converted
     * @param currency     The currency to be converted to as a 3 letter acronym, eg USD, GBP
     * @param timeInSeconds The time at which to get the price, in seconds since epoch
     * @return A double value, which <b>is not</b> rounded to any significant figures
     */
    fun getBtcHistoricPrice(
            satoshis: Long,
            currency: String,
            timeInSeconds: Long
    ): Observable<Double> = rxPinning.call<Double> {
        priceApi.getHistoricPrice(CryptoCurrencies.BTC.symbol, currency, timeInSeconds)
                .map {
                    val exchangeRate = BigDecimal.valueOf(it)
                    val satoshiDecimal = BigDecimal.valueOf(satoshis)
                    return@map exchangeRate.multiply(satoshiDecimal.divide(SATOSHIS_PER_BITCOIN))
                            .toDouble()
                }
                .compose(RxUtil.applySchedulersToObservable())
    }

    /**
     * Returns the historic value of a number of Wei at a given time in a given currency.
     *
     * @param wei          The amount of Ether to be converted in Wei, ie ETH * 1e18
     * @param currency     The currency to be converted to as a 3 letter acronym, eg USD, GBP
     * @param timeInSeconds The time at which to get the price, in seconds since epoch
     * @return A double value, which <b>is not</b> rounded to any significant figures
     */
    fun getEthHistoricPrice(
            wei: BigInteger,
            currency: String,
            timeInSeconds: Long
    ): Observable<Double> = rxPinning.call<Double> {
        priceApi.getHistoricPrice(CryptoCurrencies.ETHER.symbol, currency, timeInSeconds)
                .map {
                    val exchangeRate = BigDecimal.valueOf(it)
                    val ethDecimal = BigDecimal(wei)
                    return@map exchangeRate.multiply(ethDecimal.divide(WEI_PER_ETHER))
                            .toDouble()
                }
                .compose(RxUtil.applySchedulersToObservable())
    }

    private fun getLastPrice(currencyName: String, cryptoCurrency: CryptoCurrencies): Double {
        val prefsKey: String
        val tickerData: Map<String, PriceDatum>?

        when (cryptoCurrency) {
            CryptoCurrencies.BTC -> {
                prefsKey = PREF_LAST_KNOWN_BTC_PRICE
                tickerData = btcTickerData
            }
            CryptoCurrencies.ETHER -> {
                prefsKey = PREF_LAST_KNOWN_ETH_PRICE
                tickerData = ethTickerData
            }
            else -> throw IllegalArgumentException("BCH is not currently supported")
        }
        var currency = currencyName
        if (currency.isEmpty()) {
            currency = "USD"
        }

        var lastPrice: Double
        val lastKnown = prefsUtil.getValue("$prefsKey$currency", "0.0").toDouble()

        if (tickerData == null) {
            lastPrice = lastKnown
        } else {
            val tickerItem = when (cryptoCurrency) {
                CryptoCurrencies.BTC -> getTickerItem(currency, btcTickerData)
                CryptoCurrencies.ETHER -> getTickerItem(currency, ethTickerData)
                else -> throw IllegalArgumentException("BCH is not currently supported")
            }

            lastPrice = when (tickerItem) {
                is Optional.Some -> tickerItem.element.price ?: 0.0
                else -> 0.0
            }

            if (lastPrice > 0.0) {
                prefsUtil.setValue("$prefsKey$currency", lastPrice.toString())
            } else {
                lastPrice = lastKnown
            }
        }

        return lastPrice
    }

    private fun getBtcTicker() = rxPinning.call<Map<String, PriceDatum>> {
        priceApi.getPriceIndexes(CryptoCurrencies.BTC.symbol)
                .doOnNext { this.btcTickerData = it.toMap() }
                .compose(RxUtil.applySchedulersToObservable())
    }

    private fun getEthTicker() = rxPinning.call<Map<String, PriceDatum>> {
        priceApi.getPriceIndexes(CryptoCurrencies.ETHER.symbol)
                .doOnNext { this.ethTickerData = it.toMap() }
                .compose(RxUtil.applySchedulersToObservable())
    }

    private fun getTickerItem(
            currencyName: String,
            tickerData: Map<String, PriceDatum>?
    ): Optional<PriceDatum> {
        val priceDatum = tickerData?.get(currencyName)
        return when {
            priceDatum != null -> Optional.Some(priceDatum)
            else -> Optional.None
        }
    }

    companion object {

        private const val PREF_LAST_KNOWN_BTC_PRICE = "LAST_KNOWN_BTC_VALUE_FOR_CURRENCY_"
        private const val PREF_LAST_KNOWN_ETH_PRICE = "LAST_KNOWN_ETH_VALUE_FOR_CURRENCY_"

        private val SATOSHIS_PER_BITCOIN = BigDecimal.valueOf(100_000_000L)
        private val WEI_PER_ETHER = BigDecimal.valueOf(1e18)

        @JvmStatic
        val instance: ExchangeRateFactory by unsafeLazy { Holder.INSTANCE }

    }
}
